test/hotspot/jtreg/serviceability/sa/sadebugd/ClhsdbAttachToDebugServer.java

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * @test
+ * @bug 8262520
+ * @summary Test clhsdb connect, detach, reattach commands
+ * @requires vm.hasSA
+ * @requires os.family != "windows"
+ * @library /test/lib
+ * @run main/othervm ClhsdbAttachToDebugServer
+ */
+
+public class ClhsdbAttachToDebugServer {
+
+    public static void main(String[] args) throws Exception {
+        SATestUtils.skipIfCannotAttach(); // throws SkippedException if attach not expected to work.
+
+        if (SATestUtils.needsPrivileges()) {
+            // This tests has issues if you try adding privileges on OSX. The debugd process cannot
+            // be killed if you do this (because it is a root process and the test is not), so the destroy()
+            // call fails to do anything, and then waitFor() will time out. If you try to manually kill it with
+            // a "sudo kill" command, that seems to work, but then leaves the LingeredApp it was
+            // attached to in a stuck state for some unknown reason, causing the stopApp() call
+            // to timeout. For that reason we don't run this test when privileges are needed. Note
+            // it does appear to run fine as root, so we still allow it to run on OSX when privileges
+            // are not required.
+            throw new SkippedException("Cannot run this test on OSX if adding privileges is required.");
+        }
+
+        System.out.println("Starting ClhsdbAttachToDebugServer test");
+
+        LingeredApp theApp = null;
+        DebugdUtils debugd = null;
+        try {
+            theApp = LingeredApp.startApp();
+            System.out.println("Started LingeredApp with pid " + theApp.getPid());
+            debugd = new DebugdUtils(null);
+            debugd.attach(theApp.getPid());
+
+            JDKToolLauncher jhsdbLauncher = JDKToolLauncher.createUsingTestJDK("jhsdb");
+            jhsdbLauncher.addToolArg("clhsdb");
+
+            Process jhsdb = (SATestUtils.createProcessBuilder(jhsdbLauncher)).start();
+            OutputAnalyzer out = new OutputAnalyzer(jhsdb);
+
+            try (PrintStream console = new PrintStream(jhsdb.getOutputStream(), true)) {
+                console.println("echo true");
+                console.println("verbose true");
+                console.println("attach localhost");
+                console.println("class java.lang.Object");
+                console.println("detach");
+                console.println("reattach");
+                console.println("class java.lang.String");
+                console.println("quit");
+            }
+
+            jhsdb.waitFor();
+            System.out.println(out.getStdout());
+            System.err.println(out.getStderr());
+
+            out.stderrShouldBeEmptyIgnoreDeprecatedWarnings();
+            out.shouldMatch("^java/lang/Object @0x[0-9a-f]+$"); // for "class java.lang.Object"
+            out.shouldMatch("^java/lang/String @0x[0-9a-f]+$"); // for "class java.lang.String"
+            out.shouldHaveExitValue(0);
```

Comment:
I think you should do more error checking here. If you look at ClhsdbLauncher you will see:
```
        // This will detect most SA failures, including during the attach.
        oa.shouldNotMatch("^sun.jvm.hotspot.debugger.DebuggerException:.*$");
        // This will detect unexpected exceptions, like NPEs and asserts, that are caught
        // by sun.jvm.hotspot.CommandProcessor.
        oa.shouldNotMatch("^Error: .*$");
```
I think you could use something similar here.

Reply:
I added them in new commit.
BTW I found out to check warnings which relate to `-Xcheck:jni` at ClhsdbLauncher. I did not add them to ClhsdbAttachToDebugServer because they are not the issue which relates to debug server. Is it ok?

Reply:
No need for the `-Xcheck:jni` checks. I added that recently because we did have a bug, but I think we get more than enough testing with it already when running the ClhsdbLauncher tests.